Upgrade SUNDIALS to v5.4.0

AnHeuermann opened the pull request:
Second try to get https://github.com/OpenModelica/OpenModelica/pull/796 done.

I merged with PR https://github.com/OpenModelica/OpenModelica/pull/6737 to try the CMake build as well. When it's merged I probably need to rebase again.
Since #6737 needs a recent CMake anyway I didn't change the SUNDIALS CMake to use an older version.

I had to manually fix up the old Cpp runtime again and the CVODE added to the C runtime.

AnHeuermann commented:
I can't get the tests for SUNDIALS with KLU to succeed:
```
# Test SUNDIALS but allow it to fail
make -C 3rdParty/sundials-5.4.0/build test test_install || echo "WARNING: SUNDIALS test failing, continuing any way"
make[5]: Entering directory '/var/lib/jenkins1/ws/OpenModelica_PR-6834/OMCompiler/3rdParty/sundials-5.4.0/build'
Running tests...
/usr/bin/ctest --force-new-ctest-process 
Test project /var/lib/jenkins1/ws/OpenModelica_PR-6834/OMCompiler/3rdParty/sundials-5.4.0/build
      Start  1: ark_analytic
 1/59 Test  #1: ark_analytic ......................................   Passed    0.00 sec
      Start  2: cvRoberts_dns
 2/59 Test  #2: cvRoberts_dns .....................................   Passed    0.00 sec
      Start  3: cvsRoberts_dns
 3/59 Test  #3: cvsRoberts_dns ....................................   Passed    0.00 sec
      Start  4: idaRoberts_dns
 4/59 Test  #4: idaRoberts_dns ....................................   Passed    0.00 sec
      Start  5: idasRoberts_dns
 5/59 Test  #5: idasRoberts_dns ...................................   Passed    0.00 sec
      Start  6: kinAnalytic_fp
 6/59 Test  #6: kinAnalytic_fp ....................................   Passed    0.00 sec
      Start  7: kinAnalytic_fp_2
 7/59 Test  #7: kinAnalytic_fp_2 ..................................   Passed    0.00 sec
      Start  8: test_nvector_serial_1000_0
 8/59 Test  #8: test_nvector_serial_1000_0 ........................   Passed    0.00 sec
      Start  9: test_nvector_serial_10000_0
 9/59 Test  #9: test_nvector_serial_10000_0 .......................   Passed    0.02 sec
      Start 10: test_nvector_manyvector_1000_100_0
10/59 Test #10: test_nvector_manyvector_1000_100_0 ................   Passed    0.01 sec
      Start 11: test_nvector_manyvector_100_1000_0
11/59 Test #11: test_nvector_manyvector_100_1000_0 ................   Passed    0.01 sec
      Start 12: test_sunmatrix_dense_100_100_0
12/59 Test #12: test_sunmatrix_dense_100_100_0 ....................   Passed    0.00 sec
      Start 13: test_sunmatrix_dense_200_1000_0
13/59 Test #13: test_sunmatrix_dense_200_1000_0 ...................   Passed    0.01 sec
      Start 14: test_sunmatrix_dense_2000_100_0
14/59 Test #14: test_sunmatrix_dense_2000_100_0 ...................   Passed    0.01 sec
      Start 15: test_sunmatrix_band_10_2_3_0
15/59 Test #15: test_sunmatrix_band_10_2_3_0 ......................   Passed    0.00 sec
      Start 16: test_sunmatrix_band_300_7_4_0
16/59 Test #16: test_sunmatrix_band_300_7_4_0 .....................   Passed    0.00 sec
      Start 17: test_sunmatrix_band_1000_8_8_0
17/59 Test #17: test_sunmatrix_band_1000_8_8_0 ....................   Passed    0.00 sec
      Start 18: test_sunmatrix_band_5000_3_20_0
18/59 Test #18: test_sunmatrix_band_5000_3_20_0 ...................   Passed    0.02 sec
      Start 19: test_sunmatrix_sparse_400_400_0_0
19/59 Test #19: test_sunmatrix_sparse_400_400_0_0 .................   Passed    0.01 sec
      Start 20: test_sunmatrix_sparse_450_450_1_0
20/59 Test #20: test_sunmatrix_sparse_450_450_1_0 .................   Passed    0.01 sec
      Start 21: test_sunmatrix_sparse_200_1000_0_0
21/59 Test #21: test_sunmatrix_sparse_200_1000_0_0 ................   Passed    0.01 sec
      Start 22: test_sunmatrix_sparse_6000_350_0_0
22/59 Test #22: test_sunmatrix_sparse_6000_350_0_0 ................   Passed    0.05 sec
      Start 23: test_sunmatrix_sparse_500_5000_1_0
23/59 Test #23: test_sunmatrix_sparse_500_5000_1_0 ................   Passed    0.06 sec
      Start 24: test_sunmatrix_sparse_4000_800_1_0
24/59 Test #24: test_sunmatrix_sparse_4000_800_1_0 ................   Passed    0.06 sec
      Start 25: test_sunlinsol_band_10_2_3_0
25/59 Test #25: test_sunlinsol_band_10_2_3_0 ......................   Passed    0.00 sec
      Start 26: test_sunlinsol_band_300_7_4_0
26/59 Test #26: test_sunlinsol_band_300_7_4_0 .....................   Passed    0.00 sec
      Start 27: test_sunlinsol_band_1000_8_8_0
27/59 Test #27: test_sunlinsol_band_1000_8_8_0 ....................   Passed    0.00 sec
      Start 28: test_sunlinsol_band_5000_3_100_0
28/59 Test #28: test_sunlinsol_band_5000_3_100_0 ..................   Passed    0.03 sec
      Start 29: test_sunlinsol_dense_10_0
29/59 Test #29: test_sunlinsol_dense_10_0 .........................   Passed    0.00 sec
      Start 30: test_sunlinsol_dense_100_0
30/59 Test #30: test_sunlinsol_dense_100_0 ........................   Passed    0.00 sec
      Start 31: test_sunlinsol_dense_500_0
31/59 Test #31: test_sunlinsol_dense_500_0 ........................   Passed    0.04 sec
      Start 32: test_sunlinsol_dense_1000_0
32/59 Test #32: test_sunlinsol_dense_1000_0 .......................   Passed    0.40 sec
      Start 33: test_sunlinsol_klu_300_0_0
33/59 Test #33: test_sunlinsol_klu_300_0_0 ........................***Failed    0.02 sec
      Start 34: test_sunlinsol_klu_300_1_0
34/59 Test #34: test_sunlinsol_klu_300_1_0 ........................***Failed    0.01 sec
      Start 35: test_sunlinsol_klu_1000_0_0
35/59 Test #35: test_sunlinsol_klu_1000_0_0 .......................***Failed    0.15 sec
      Start 36: test_sunlinsol_klu_1000_1_0
36/59 Test #36: test_sunlinsol_klu_1000_1_0 .......................***Failed    0.14 sec
      Start 37: test_sunlinsol_lapackband_10_2_3_0
37/59 Test #37: test_sunlinsol_lapackband_10_2_3_0 ................   Passed    0.02 sec
      Start 38: test_sunlinsol_lapackband_300_7_4_0
38/59 Test #38: test_sunlinsol_lapackband_300_7_4_0 ...............   Passed    0.00 sec
      Start 39: test_sunlinsol_lapackband_1000_8_8_0
39/59 Test #39: test_sunlinsol_lapackband_1000_8_8_0 ..............   Passed    0.00 sec
      Start 40: test_sunlinsol_lapackband_5000_3_100_0
40/59 Test #40: test_sunlinsol_lapackband_5000_3_100_0 ............   Passed    0.02 sec
      Start 41: test_sunlinsol_lapackdense_10_0
41/59 Test #41: test_sunlinsol_lapackdense_10_0 ...................   Passed    0.00 sec
      Start 42: test_sunlinsol_lapackdense_100_0
42/59 Test #42: test_sunlinsol_lapackdense_100_0 ..................   Passed    0.00 sec
      Start 43: test_sunlinsol_lapackdense_500_0
43/59 Test #43: test_sunlinsol_lapackdense_500_0 ..................   Passed    0.03 sec
      Start 44: test_sunlinsol_lapackdense_1000_0
44/59 Test #44: test_sunlinsol_lapackdense_1000_0 .................   Passed    0.19 sec
      Start 45: test_sunlinsol_spgmr_serial_100_1_1_100_1e-13_0
45/59 Test #45: test_sunlinsol_spgmr_serial_100_1_1_100_1e-13_0 ...   Passed    0.00 sec
      Start 46: test_sunlinsol_spgmr_serial_100_2_1_100_1e-13_0
46/59 Test #46: test_sunlinsol_spgmr_serial_100_2_1_100_1e-13_0 ...   Passed    0.00 sec
      Start 47: test_sunlinsol_spgmr_serial_100_1_2_100_1e-13_0
47/59 Test #47: test_sunlinsol_spgmr_serial_100_1_2_100_1e-13_0 ...   Passed    0.00 sec
      Start 48: test_sunlinsol_spgmr_serial_100_2_2_100_1e-13_0
48/59 Test #48: test_sunlinsol_spgmr_serial_100_2_2_100_1e-13_0 ...   Passed    0.00 sec
      Start 49: test_sunlinsol_spfgmr_serial_100_1_100_1e-13_0
49/59 Test #49: test_sunlinsol_spfgmr_serial_100_1_100_1e-13_0 ....   Passed    0.00 sec
      Start 50: test_sunlinsol_spfgmr_serial_100_2_100_1e-13_0
50/59 Test #50: test_sunlinsol_spfgmr_serial_100_2_100_1e-13_0 ....   Passed    0.00 sec
      Start 51: test_sunlinsol_spbcgs_serial_100_1_100_1e-13_0
51/59 Test #51: test_sunlinsol_spbcgs_serial_100_1_100_1e-13_0 ....   Passed    0.00 sec
      Start 52: test_sunlinsol_spbcgs_serial_100_2_100_1e-13_0
52/59 Test #52: test_sunlinsol_spbcgs_serial_100_2_100_1e-13_0 ....   Passed    0.00 sec
      Start 53: test_sunlinsol_sptfqmr_serial_100_1_100_1e-13_0
53/59 Test #53: test_sunlinsol_sptfqmr_serial_100_1_100_1e-13_0 ...   Passed    0.00 sec
      Start 54: test_sunlinsol_sptfqmr_serial_100_2_100_1e-13_0
54/59 Test #54: test_sunlinsol_sptfqmr_serial_100_2_100_1e-13_0 ...   Passed    0.00 sec
      Start 55: test_sunlinsol_pcg_serial_100_500_1e-13_0
55/59 Test #55: test_sunlinsol_pcg_serial_100_500_1e-13_0 .........   Passed    0.00 sec
      Start 56: test_sunnonlinsol_newton
56/59 Test #56: test_sunnonlinsol_newton ..........................   Passed    0.00 sec
      Start 57: test_sunnonlinsol_fixedpoint
57/59 Test #57: test_sunnonlinsol_fixedpoint ......................   Passed    0.00 sec
      Start 58: test_sunnonlinsol_fixedpoint_2
58/59 Test #58: test_sunnonlinsol_fixedpoint_2 ....................   Passed    0.00 sec
      Start 59: test_sunnonlinsol_fixedpoint_2_0.5
59/59 Test #59: test_sunnonlinsol_fixedpoint_2_0.5 ................   Passed    0.00 sec

93% tests passed, 4 tests failed out of 59

Total Test time (real) =   1.49 sec

The following tests FAILED:
	 33 - test_sunlinsol_klu_300_0_0 (Failed)
	 34 - test_sunlinsol_klu_300_1_0 (Failed)
	 35 - test_sunlinsol_klu_1000_0_0 (Failed)
	 36 - test_sunlinsol_klu_1000_1_0 (Failed)
Errors while running CTest
Makefile:121: recipe for target 'test' failed
make[5]: *** [test] Error 8
make[5]: Leaving directory '/var/lib/jenkins1/ws/OpenModelica_PR-6834/OMCompiler/3rdParty/sundials-5.4.0/build'
WARNING: SUNDIALS test failing, continuing any way
```

It doesn't work on my machine or on Jenkins. Could be that the test is broken, because all tests in the testsuite are passing on my machine (or we just don't test SUNDIALS with KLU...).

AnHeuermann commented:
I fixed the MINGW and MSVC builds, but the `all-runtimes` target in `Makeflie.omdev.mingw` is giving me a hard time.
The Problem is basically, that we compile 3rdParty libraries over and over again and out of order.
`sundials_msvc` needs `umfpack_msvc` build first because of klu.

I [added a check](https://github.com/OpenModelica/OpenModelica/pull/6834/commits/3737f8047500431ebf64e86a3f99536a47b4f86a#diff-a26b3be37351da12715b1803fb766c55a3bd7b60c2722499f1d73597cefb2580R432) in `sundials_msvc` to see if `umfpack_msvc` finished and copied `klu.lib` to the build folder.
The error message from Jenkins:
```
test -f "C:/dev/jenkins/ws/OpenModelica_PR-6834/build/lib/omc/msvc/klu.lib"
echo @echo on >> 3rdParty/CMinpack/build_msvc/build.bat
echo @echo on >> 3rdParty/FMIL/build_msvc/build.bat
Makefile.omdev.mingw:423: recipe for target '/c/dev/jenkins/ws/OpenModelica_PR-6834/build/lib/omc/msvc/sundials_ida.dll' failed
make[1]: *** [/c/dev/jenkins/ws/OpenModelica_PR-6834/build/lib/omc/msvc/sundials_ida.dll] Error 1
make[1]: *** Waiting for unfinished jobs....
```
If I read the log from Jenkins correctly the `umfpack_msvc` target was not started yet, but `sundials_msvc` was.

Make is ignoring my dependencies I added to `sundials_msvc` so I tried to do stuff seriell with https://github.com/OpenModelica/OpenModelica/pull/6834/commits/a728da04026eb8ab84b715deee4930588a1d942b and
don't run different build types (MSVC and MINGW) in parallel:
```Makefile
all-runtimes:
	$(MAKE) -C OMCompiler -f $(defaultMakefileTarget) OMBUILDDIR=$(OMBUILDDIR) BUILDTYPE=$(BUILDTYPE) runtimeCPPinstall
	$(MAKE) -C OMCompiler -f $(defaultMakefileTarget) OMBUILDDIR=$(OMBUILDDIR) BUILDTYPE=$(BUILDTYPE) runtimeCPPmsvcinstall
	$(MAKE) -C OMCompiler -f $(defaultMakefileTarget) OMBUILDDIR=$(OMBUILDDIR) BUILDTYPE=$(BUILDTYPE) simulationruntimecmsvc
```

@adrpo and @mahge Any ideas how to get this running?
Also the new OMDev version will probably change stuff, but hopfully only makes it better.

adrpo commented:
How did you put the umfpack_msvc as a dependency for sundials_msvc?

adrpo commented:
When I fix the broken msys2/mingw compilation using #6872 I will have a look at this one.

AnHeuermann commented:
I added
```Makefile
sundials_msvc: $(OMBUILDDIR)/lib/omc/msvc/sundials_ida.dll
$(OMBUILDDIR)/lib/omc/msvc/sundials_ida.dll: getMSVCversion 3rdParty/sundials-5.4.0/CMakeLists.txt Umfpack_msvc
[...]
```
to OMCompiler/Makefile.omdev.mingw and Make started `sundials_msvc` bevore `Umfpack_msvc` was finished.

mahge commented on testsuite/simulation/modelica/solver/problem6-cvode.mos:
```diff
@@ -83,7 +83,7 @@ if res2[1,s] > -1e-2 then 1 else 0;
 // true
 // 1.0
-// 1.0
+// 0.0
```
I guess this was wrong the whole time and it is now correct. That is a good sign.

AnHeuermann replied:
Yes. I'm not sure why I didn't noticed at first, but now it should be correct.

mahge commented on OMCompiler/m4/lapack.m4:
```diff
@@ -22,6 +23,7 @@ AC_DEFUN([OMC_AC_LAPACK], [
     LIBS=""
 
+    # Auto detect using pkg-config
     if test "$LD_LAPACK" = "auto"; then
```
This indentation might be wrong.

AnHeuermann replied:
It looks correct to me. I wanted to describe the if-case.
Or what exectly do ypu mean?

mahge replied:
Never-mind, my bad. I was looking at the github collapsed code and it seemed the `if ` at line `8` was closed by the done at `22`. The for loop was hidden by the collapsing.

mahge commented on OMCompiler/SimulationRuntime/c/simulation/solver/linearSolverKlu.h:
```diff
@@ -38,8 +38,8 @@
 #include "../../simulation_data.h"
-#include "suitesparse/Include/amd.h"
-#include "suitesparse/Include/klu.h"
+#include "amd.h"
```
"" kind of includes are usually indicate the header is part of you local project or code. If headers are from an outside library through an include directive you can use <> to signify that. This is of course just convention and only helps for overlook feel at a glance.

AnHeuermann replied:
Yeah, you are right. I changed it at some places and will do it here as well.

mahge commented on OMCompiler/SimulationRuntime/OMSICpp/CMakeLists.txt:
```diff
@@ -921,15 +917,15 @@ IF(ZeroMQ_FOUND AND cppzmq_FOUND)
+    message (STATUS "cppzmq version ${cppzmq_VERSION}")
+    message (STATUS "Using ZMQ Header ${ZeroMQ_INCLUDE_DIR}" )
+    message (STATUS "Using CPPZMQ Header ${cppzmq_INCLUDE_DIR}" )
+ELSE(ZeroMQ_found AND cppzmq_found)
```
Why are these lowercase. CMake won't care but if it signifies that a package is found it is usually _FOUND.

mahge reviewed (approved):
Overall LGTM!

There are some TODOs which seem to have been fixed already but the comments are still left.
These and issues with individual comments can be fixed afterwards. 

The testsuite failures seem to be white-space issues mostly introduced by yourself.

You have also done a lot of formatting fixes that you did not technically have to do for this PR. That's good!

adrpo commented on OMCompiler/SimulationRuntime/c/simulation/solver/kinsolSolver.h:
```diff
@@ -1,7 +1,7 @@
  * This file is part of OpenModelica.
  *
- * Copyright (c) 1998-CurrentYear, Open Source Modelica Consortium (OSMC),
+ * Copyright (c) 1998-2014, Open Source Modelica Consortium (OSMC),
```
Should be CurrentYear

adrpo reviewed (approved):
Ok with me, just fix the year.
I'm not really liking the 5.4.0 in the directory name but it doesn't matter.

AnHeuermann commented:
We are very inconsistent with this. Also I have no clue how the legal situation is here.
I'll change it and merge after that.

casella commented:
@AnHeurmann, please check the latest [testuite report](https://libraries.openmodelica.org/branches/history/newInst/2020-11-09%2020:32:27..2020-11-10%2020:33:05.html). For example, something changed in the results of the ScalableTestSuite examples that causes them to fail verification. If you look at the actual [failing traces](https://libraries.openmodelica.org/branches/newInst/ScalableTestSuite/files/ScalableTestSuite_ScalableTestSuite.Thermal.Advection.ScaledExperiments.AdvectionReaction_N_6400.diff.html) it seems that we are trading some numerical noise for some different numerical noise. But at least it is clear that something changed.

Any idea?

AnHeuermann commented:
I'm currently fixing the build on RedHeat with CMake 2.8. Directly after that I will lock into this in more detail.

It is not unlikely that I introduced some new bugs with my commit.
But of course I would expect some numeric changes after such a big update (version 2.6.2 was from 2015). I updated our main solvers IDA and CVODE as well as our non-linear solver KINSOL.

So if one of those is involved in errors it could be very well be that there is some bug. And of course IDA is used nearly everywhere. Did you find a case where it really looks line an error? I find it challenging to read the reports.

AnHeuermann commented:
According to @perost the regressions are already fixed.

perost commented:
> According to @perost the regressions are already fixed.

Sorry, I didn't mean those regressions, I meant [these ones](https://libraries.openmodelica.org/branches/history/newInst/2020-11-11%2011:33:59..2020-11-11%2020:35:27.html). The regressions you're talking about here were almost certainly due to the SUNDIALS update.

AnHeuermann commented:
Ahh okay, my bad. Had me wondering already.

casella commented:
> But of course I would expect some numeric changes after such a big update (version 2.6.2 was from 2015). I updated our main solvers IDA and CVODE as well as our non-linear solver KINSOL.

OK
 
> So if one of those is involved in errors it could be very well be that there is some bug. And of course IDA is used nearly everywhere.

Is it? In ticket [https://trac.openmodelica.org/OpenModelica/ticket/4990](#4990) I proposed to switch from DASSL (which is quite old and no longer maintained) to IDA (which it its heir) as the default solver. The discussion has languished for a while, maybe we should restart it for 1.18.0 (I guess it's too late for 1.17.0). Before that, I assume that DASSL is use for most of the C runtime, except `daeMode` tests that use it by default. Judging from the [report](https://libraries.openmodelica.org/branches/history/newInst/2020-11-09%2020:32:27..2020-11-10%2020:33:05.html), the simulation logs of many models using the cpp runtime show failures, so I guess the standard solver is already IDA for that runtime (but not for the C runtime), and apparently the 5.4.0 commit had some consequences.

I thought ScalableTestSuite was using DASSL (that's what happens in OMEdit), but indeed those tests are carried out with IDA as it can be seen from the simulation logs, e.g.[ this one](https://libraries.openmodelica.org/branches/newInst/ScalableTestSuite/files/ScalableTestSuite_ScalableTestSuite.Electrical.TransmissionLine.ScaledExperiments.TransmissionLineEquations_N_10.sim). Probably we added `-s=ida` to the testsuite setup, @sjoelund do you remember why?

That explains why your commit affected them, which puzzled me at first.

If you look at the plots of the verification failures, that are available from the [library report](https://libraries.openmodelica.org/branches/newInst/ScalableTestSuite/ScalableTestSuite.html), they all involve derivatives, which are not subject to error control (the states are!), and thus tend to be noisy and have large relative errors, because derivatives are (very) close to zero when getting near equilibrium. Of course if you change version of IDA you still get noise, just a different one. Hence the verification "failures", which are actually not failures at all.

I guess I should just remove all the derivatives from the reference trajectories, because they are inevitably going to generate false negatives.

Conclusion: ScalableTestSuite regressions are not regressions at all. Other regressions involving the cpp runtime include simulation failures, those should probably be addressed.

sjoelund commented:
> Probably we added -s=ida to the testsuite setup, @sjoelund do you remember why?

The library can choose the solver:
https://github.com/casella/ScalableTestSuite/blob/master/ScalableTestSuite/Electrical/TransmissionLine.mo#L313

sjoelund commented:
> In ticket https://trac.openmodelica.org/OpenModelica/ticket/4990 I proposed to switch from DASSL (which is quite old and no longer maintained) to IDA (which it its heir) as the default solver.

Why not cvode? We don't really use dassl as a DAE solver...

AnHeuermann commented:
@casella I definitely broke both Cpp runtimes (SimulationRuntime/cpp and SimulationRuntime/OMSICpp) with different problems.
I'll come up with a fix for that. Shouldn't be that hard, but debugging the Cpp runtime is no fun.
Also having two different versions of Cpp runtime at the same time is complicating it further.

casella commented:
> @casella I definitely broke both Cpp runtimes (SimulationRuntime/cpp and SimulationRuntime/OMSICpp) with different problems.
> I'll come up with a fix for that. Shouldn't be that hard, but debugging the Cpp runtime is no fun.

Sorry about that.

> Also having two different versions of Cpp runtime at the same time is complicating it further.

The old cpp is there as an interim solution. We should push forward the upgrade of the OMSICpp runtime util it is acceptable for ABB, so the old cpp runtime can be retired. Otherwise we are going exactly in the wrong direction (more runtimes, instead of less). Please remind me that in the next devmeeting on Monday

AnHeuermann commented:
The problem goes a bit deeper than I thought, I opened a ticket about it: https://trac.openmodelica.org/OpenModelica/ticket/6228
